File: backend/app/schemas/payment.py
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class PaymentOrderResponse(BaseModel):

    payment_id: int
    razorpay_order_id: str
    amount: Decimal
    amount_in_paise: int
    currency: str
    razorpay_key_id: str

class PaymentVerifyRequest(BaseModel):
    razorpay_payment_id: str
    razorpay_order_id: str
    razorpay_signature: str

class PaymentResponse(BaseModel):
    id: int
    booking_id: int
    razorpay_order_id: str
    razorpay_payment_id: str | None
    amount: Decimal
    currency: str
    status: str

    model_config = ConfigDict(
        from_attributes=True
    )

File: frontend/src/app/core/services/payment/payment.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

export interface PaymentOrderResponse {
  payment_id: number;
  razorpay_order_id: string;
  amount: number;
  amount_in_paise: number;
  currency: string;
  razorpay_key_id: string;
}

export interface PaymentVerifyRequest {
  razorpay_payment_id: string;
  razorpay_order_id: string;
  razorpay_signature: string;
}

export interface PaymentResponse {
  id: number;
  booking_id: number;
  razorpay_order_id: string;
  razorpay_payment_id: string | null;
  amount: number;
  currency: string;
  status: string;
}

@Injectable({
  providedIn: 'root',
})
export class PaymentService {
  private readonly http = inject(HttpClient);

  private readonly API_URL =
    'http://localhost:8000/api/v1/payments';

  createOrder(
    bookingId: number
  ): Observable<PaymentOrderResponse> {
    return this.http.post<PaymentOrderResponse>(
      `${this.API_URL}/create-order/${bookingId}`,
      {}
    );
  }

  verifyPayment(
    paymentData: PaymentVerifyRequest
  ): Observable<PaymentResponse> {
    return this.http.post<PaymentResponse>(
      `${this.API_URL}/verify`,
      paymentData
    );
  }
}